src/components/collaborative-editing/diff/getChangesFromString.ts

```diff
@@ -0,0 +1,153 @@
+/**
+ * Possible operation types
+ */
+export enum Operation {
+  /**
+   * User inserts symbols
+   */
+  Insert,
+
+  /**
+   * User removes symbols
+   */
+  Remove,
+
+  /**
+   * User replaces symbols
+   */
+  Replace
+}
+
+type InsertOperation = {
+  type: Operation.Insert;
+
+  /**
+   * Index where user inserted symbols
+   */
+  from: number;
+
+  /**
+   * Which symbols user inserted
+   */
+  data: string;
+
+  /**
+   * Length of inserted data
+   */
+  length: number;
+}
+
+type RemoveOperation = {
+  type: Operation.Remove;
+
+  /**
+   * Index where user removed symbols
+   */
+  from: number;
+
+  /**
+   * Which symbols user removed
+   */
+  data: string;
+
+  /**
+   * Length of removed data
+   */
+  length: number;
+}
+
+type ReplaceOperation = {
+  type: Operation.Replace;
+
+  /**
+   * Index where user replaced symbols
+   */
+  from: number;
+  data: {
+    /**
+     * Replaced data
+     */
+    before: string;
+
+    /**
+     * New data
+     */
+    after: string;
+  };
+  length: {
+    /**
+     * Length of replaced data
+     */
+    before: number;
+
+    /**
+     * Length of new data
+     */
+    after: number;
+  };
+}
+
+/**
+ * Function finds difference between two different strings
+ *
+ * @param str1 - original string
+ * @param str2 - new string that comes from the first
+ *
+ * @returns {object} object of operation
+ */
+export function getChangesFromString(str1: string, str2: string): InsertOperation | RemoveOperation | ReplaceOperation {
+  const length1 = str1.length,
+      length2 = str2.length;
+
+  let left = 0;
+
+  while (str1[left] === str2[left]) {
```

Comment:
Please provide comments for the main logical blocks.

Reply:
In case of equal strings this cycle would go infinite